### PR TITLE
Add -blocksonly option

### PR DESCRIPTION
**Pull request by pstratem**
Set fRelay=false and reject all transactions into the mempool

### CODE REVIEW

**Comment by dcousens** [comment]
Concept ACK

**Comment by luke-jr** [comment]
Should this force wallet off? But I suppose maybe watch-only wallets don't need to submit txns...

**Comment by gmaxwell** [comment]
@luke-jr Nah!  And, in fact with this set you can still submit txns (though it'll be even more obvious its yours!)

Concept Ack. Will test.

**Comment by luke-jr** [comment]
As the code is right now, I don't think you can submit txns...

**Comment by gmaxwell** [comment]
@luke-jr Yup see my comment. :)

**Comment by pstratem** [comment]
@gmaxwell and ruin my record for smallest patch to add a new feature? :)

**Comment by gmaxwell** [comment]
Perhaps there is a reason it was record setting? :P

**Comment by gmaxwell** [comment]
If, instead of blocking at the mempool it did not getdata, then local transactions would work, and you just need to disable the getdata inhibition for whitebinded peers, and the peer is local-- so you'd keep your same lines of code count. Happy?

**Comment by paveljanik** [comment]
The option should be documented, so yet another line is needed :-)

**Comment by paveljanik** [comment]
What will we do to remote peers relaying transactions to us ignoring we announced that we do not want transactions in `version` message?

**Comment by laanwj** [comment]
Concept ACK.

> Should this force wallet off? But I suppose maybe watch-only wallets don't need to submit txns...

Otherwise there's already `-walletbroadcast=0`. To run a wallet without broadcasting.

**Comment by jonasschnelli** [comment]
Concept ACK.
Agree with @paveljanik: needs a help message line and maybe a comment on src/net.cpp:L454.

**Comment by MarcoFalke** [comment]
> Otherwise there's already -walletbroadcast=0

Use parameter interaction and set `-walletbroadcast=0` with `-blocksonly`?

**Comment by laanwj** [comment]
@MarcoFalke Yes.

There was another similar issue raised on IRC: transactions from whitelisted peers. Even though nothing is accepted into the mempool, these are still relayed:

https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L4462

File: src/main.cpp [hunk 1]
```diff
@@ -4451,7 +4451,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             assert(recentRejects);
             recentRejects->insert(tx.GetHash());
 
-            if (pfrom->fWhitelisted) {
+            if (pfrom->fWhitelisted && GetBoolArg("-whitelistalwaysrelay", true)) {
```
Comment (laanwj) [inline src/main.cpp]: Please define a `static const bool DEFAULT_WHITELISTALWAYSRELAY = true;` in `main.h`  and use that
(sorry for making your patch even more hideously huge)

Comment (MarcoFalke) [inline src/main.cpp]: Also, HelpMessage needs update?

**Comment by pstratem** [comment]
@MarcoFalke I'm not sure we cant to put either of these options in the HelpMessage

Unless you know what you're doing these are really not a good idea to change.

File: src/main.cpp (hunk 1 above)
Comment (MarcoFalke) [inline src/main.cpp]: @pstratem You can hide them behind `showDebug`.

**Comment by jonasschnelli** [comment]
@pstratem: for the sake of a central startup arguments "documentation" I think we need to add a help message for every startup parameter. If it's a debugging/expert only feature, it should be placed behind the `-debughelp` wall.

**Comment by petertodd** [comment]
We really should define something like NODE_BLOCKS and NODE_TXS... Doesn't necessarily need to be in this pull, but worth noting.

**Comment by pstratem** [comment]
@petertodd agreed

**Comment by gmaxwell** [comment]
Needs rebase.   And my tested/code ack is held only on the fact that there is a global static setting for the default thats not used everywhere.  :(

**Comment by laanwj** [comment]
There was also talk about sending an empty filterload to peers in this mode, if they support NODE_BLOOM.
Let's do this (if still relevant) in a later pull, just mentioning it for completeness.

**Comment by gmaxwell** [comment]
@laanwj We don't need to send an empty filterload because the relay flag has the same effect!

**Comment by gmaxwell** [comment]
@pstratem Move the DEFAULT_BLOCKSONLY to net.h, then you can use it everywhere uniformly and future programmers will not hate you quite so much.  (thanks to @laanwj for asking the dumb question I missed. :) )

**Comment by dcousens** [comment]
Needs rebase

**Comment by rebroad** [comment]
Erk.. This is really ugly... And the pull doesn't even state who needs this functionality or why.

**Comment by luke-jr** [comment]
@rebroad Basically this makes Core a blockchain-only node and disables off-chain transactions.

**Comment by gmaxwell** [comment]
@rebroad It's fairly self explanatory. If you want to run a node that only processes blocks (e.g. don't care about unconfirmed transactions); you can set this and save ~80% of your bandwidth, including all redundant transmission of block data; "Blocksonly". This is good for many applications (including my laptop... :) )

**Comment by rebroad** [comment]
Ah, I thought I had raised a pull request for my Antisocial fork, but it seems not. I think it would be useful to have an option to do blocksonly during IBD only.  Would this be of interest to anyone?

**Comment by laanwj** [comment]
> to do blocksonly during IBD only

@rebroad That should already be the case?

**Pull request merged**